cmd/snap/cmd_run.go

```diff
@@ -1013,3 +1062,256 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
+				return errDBusUnknownMethod
+			case "org.freedesktop.DBus.Error.Spawn.ChildExited":
+				// We tried to socket-activate dbus-daemon or bus-activate
+				// systemd --user but it failed.
+				return errDBusSpawnChildExited
+			case "org.freedesktop.systemd1.UnitExists":
+				// Starting a scope with a name that already exists is an
+				// error. Normally this should never happen.
+				return fmt.Errorf("cannot create transient scope: scope %q clashed: %s", unitName, err)
+			default:
+				return fmt.Errorf("cannot create transient scope: DBus error %q: %v", dbusErr.Name, dbusErr.Body)
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("cannot create transient scope: %s", err)
+		}
+	}
+	logger.Debugf("created transient scope as object: %s", job)
+	return nil
+}
+
+var dbusSystemBus = func() (*dbus.Conn, error) {
+	return dbus.SystemBus()
+}
+
+var osGetuid = os.Getuid
+
+// isSessionBusLikelyPresent checks for the apparent availability of DBus session bus.
+//
+// The code matches what go-dbus does when it tries to detect the session bus:
+// - the presence of the environment variable DBUS_SESSION_BUS_ADDRESS
+// - the presence of the bus socket address in the file /run/user/UID/dbus-session
+// - the presence of the bus socket in /run/user/UID/bus
+func isSessionBusLikelyPresent() bool {
+	if address := os.Getenv("DBUS_SESSION_BUS_ADDRESS"); address != "" {
+		return true
+	}
+	if fi, err := os.Stat(fmt.Sprintf("/run/user/%d/dbus-session", osGetuid())); err == nil {
+		if stat, ok := fi.Sys().(*syscall.Stat_t); ok {
+			if stat.Mode&syscall.S_IFREG == syscall.S_IFREG {
+				return true
+			}
+		}
+	}
+	if fi, err := os.Stat(fmt.Sprintf("/run/user/%d/bus", osGetuid())); err == nil {
+		if stat, ok := fi.Sys().(*syscall.Stat_t); ok {
+			if stat.Mode&syscall.S_IFSOCK == syscall.S_IFSOCK {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+var dbusSessionBus = func() (*dbus.Conn, error) {
+	// This started as a helper for mocking and testing but we ended up needing
+	// a custom implementation to prevent go-dbus from spawning and *leaving
+	// behind* a dbus-launch process that will never be used by anything again.
+	if isSessionBusLikelyPresent() {
+		return dbus.SessionBus()
+	}
+	return nil, fmt.Errorf("cannot find session bus")
+}
+
+var randomUUID = func() (string, error) {
+	// The source of the bytes generated here is the same as that of
+	// /dev/urandom which doesn't block and is sufficient for our purposes
+	// of avoiding clashing UUIDs that are needed for all of the non-service
+	// commands that are started with the help of this UUID.
+	uuidBytes, err := ioutil.ReadFile("/proc/sys/kernel/random/uuid")
+	return strings.TrimSpace(string(uuidBytes)), err
+}
+
+var osGetpid = os.Getpid
+
+var cgroupProcessPathInTrackingCgroup = cgroup.ProcessPathInTrackingCgroup
+
+var errCannotTrackProcess = errors.New("cannot track application process")
+
+var createTransientScope = func(securityTag string) error {
```

Comment:
it's a bit unclear where this function should live, it is used here, it's very tied to sandbox cgroup though, otoh it would create a dep between that and systemd, maybe that's not a problem

Reply:
It's now moved to the sandbox/cgroup package